### ISSUE TITLE
Async csv read stream using csv-core and tokio

### ISSUE THREAD

**Issue by dgllghr** [BurntSushi/rust-csv#141]
#### What version of the `csv` crate are you using?

csv-core 0.1.5

#### Briefly describe the question, bug or feature request.

I wrote an implementation of an async streaming csv parser using csv-core and tokio as part of a larger project. I've included the code for the csv read stream below. If there is interest in including it in this project, I can open a pull request. If not, I will try to put together a more complete library based on csv-core and publish it independently to crates.io.

#### Include a complete program demonstrating a problem.

https://gist.github.com/dgllghr/2b93d28a124f65d26d5b500a21789151

#### What is the observed behavior of the code above?

Works for the use cases I have thrown at it but has not been vigorously tested

**Comment by BurntSushi** [comment]
Interesting, thank you!

I would certainly like this crate to support async, but I'm not ready to start that process until there is a clear ecosystem wide rallying point. My actual hope is that there will exist a generic adapter that causes things that implement `std::io::Read` to also implement `AsyncRead` so that you don't need to hand-roll two implementations for everything. Hmmm, actually, it looks like [that already exists](https://docs.rs/tokio-io/0.1.11/tokio_io/io/struct.AllowStdIo.html). It is possible to use `AllowStdIo` in your case?

**Comment by dgllghr** [comment]
I wasn't aware of `AllowStdIo`, but it looks like it will block the executor if used with underlying IO readers that block. One thing that might work would be to pass the underlying async stream to the `csv::Reader` (because `AsyncRead: std::io::Read`) , wrap the `csv::Reader` in `AllowStdIo`, and then use the wrapped reader in an asynchronous context. I am not totally sure there won't be any blocking, though.

Either way, I agree that being able to use the same IO code synchronously and asynchronously would be a huge boon and I understand your position about incorporating the async code. I'll try to spend some time in the next few days to test to see if `AllowStdIo` would work and not block.

**Comment by BurntSushi** [comment]
Thanks so much @dgllghr! That looks promising. For now, I'm going to close this since I don't think there's anything actionable here for the `csv` crate. Once the async ecosystem stabilizes, we can revisit this if necessary. My hope though is that there's nothing that the `csv` crate needs to do.

**Comment by dgllghr** [comment]
@BurntSushi, I know you decided not to pursue async in this crate at the moment, but I wanted to post an issue I found with the code I posted earlier. Unfortunately, the reader has to be aware of the `WouldBlock` error in order to manage state internally correctly. Because the `csv::Reader` is not aware of `WouldBlock`, it is possible to get into scenarios where hitting `WouldBlock` at the wrong time causes the internal state of the reader to be out of sync with the underlying reader. This isn't an issue with `csv::Reader`, it just highlights that async compatibility requires more work than just a compatibility shim. I wanted to post this follup-up for completeness sake in case someone stumbled across the previous code and thought it was safe for use.
